src/iotHubMessageExplorer.ts

```diff
@@ -21,27 +69,97 @@ export class IoTHubMessageExplorer extends IoTHubMessageBaseExplorer {
             if (message !== undefined) {
                 this._outputChannel.show();
                 try {
                     let client = clientFromConnectionString(deviceConnectionString);
-                    let stringify = Utility.getConfig<boolean>(Constants.IoTHubD2CMessageStringifyKey);
-                    client.sendEvent(new Message(stringify ? JSON.stringify(message) : message),
-                        this.sendEventDone(client, Constants.IoTHubMessageLabel, Constants.IoTHub, Constants.IoTHubAIMessageDoneEvent));
+                    this.sendD2CMessageCore(client, message);
                 } catch (e) {
                     this.outputLine(Constants.IoTHubMessageLabel, e);
                 }
             }
         });
     }
 
+    private async sendD2CMessageCore(client: Client, message: string) {
+        let stringify = Utility.getConfig<boolean>(Constants.IoTHubD2CMessageStringifyKey);
+        client.sendEvent(new Message(stringify ? JSON.stringify(message) : message),
+            this.sendEventDone(client, Constants.IoTHubMessageLabel, Constants.IoTHub, Constants.IoTHubAIMessageDoneEvent));
+    }
+
+    private async sendD2CMessageCoreWithProgress(client: Client, message: string, status: SendStatus, totalStatus: SendStatus) {
+        let stringify = Utility.getConfig<boolean>(Constants.IoTHubD2CMessageStringifyKey);
+        await client.sendEvent(new Message(stringify ? JSON.stringify(message) : message),
+            this.sendEventDoneWithProgress(client, Constants.IoTHubAIMessageDoneEvent, status, totalStatus));
+    }
+
+    private async delay(ms: number, token?: vscode.CancellationToken) {
+        if (ms <= 1000) {
+            return new Promise( resolve => setTimeout(resolve, ms));
+        } else {
+            await new Promise( resolve => setTimeout(resolve, 1000));
+            if (token) {
+                if (token.isCancellationRequested) {
+                    return;
+                } else {
+                    await this.delay(ms - 1000, token);
+                }
+            } else {
+                await this.delay(ms - 1000);
+            }
+        }
+    }
+
+    public async sendD2CMessageFromMultipleDevicesRepeatedlyWithProgressBar(deviceConnectionStrings: string[], message: string, times: number, interval: number) {
+        const startTime = new Date();
+        this.outputLine(Constants.SimulatorSummaryLabel, `Start at ${startTime}`)
+        await vscode.window.withProgress({
+			location: vscode.ProgressLocation.Notification,
+			title: Constants.SimulatorSendingMessageProgressBarTitle,
+			cancellable: true
+		}, async (progress, token) => {
+            token.onCancellationRequested(() => {
+                vscode.window.showInformationMessage(Constants.SimulatorProgressBarCancelLog);
+            })
+            progress.report({ increment: 0});
+            const deviceCount = deviceConnectionStrings.length;
+            const total = deviceCount * times;

```

Comment:
Any check to guard total!=0?

Reply:
Since it guarantees the deviceConnectionStrings(an array) will not be empty, and the input validator in webview will make sure it is a positive integer(greater than 0, and is not float), so total will never be <=0. But to make it safer, I will add additional check here.

Reply:
Modified.